lib/common/decorators.ts

```diff
@@ -83,3 +85,48 @@ export function exported(moduleName: string): any {
 		return descriptor;
 	};
 }
+
+export function performanceLog(injector?: IInjector): any {
+	injector = injector || $injector;
+	return function (target: any, propertyKey: string, descriptor: PropertyDescriptor): any {
+		const originalMethod = descriptor.value;
+		const className = target.constructor.name;
+		const trackName = `${className}${AnalyticsEventLabelDelimiter}${propertyKey}`;
+		const performanceService: IPerformanceService = injector.resolve("performanceService");
+
+		//needed for the returned function to have the same name as the original - used in hooks decorator
+		const tempObject = {
+			[originalMethod.name]: function (...args: Array<any>) {
+				const start = performanceService.now();
+				const result = originalMethod.apply(this, args);
+				const resolvedPromise = Promise.resolve(result);
+				let end;
+
+				if (resolvedPromise !== result) {
+					end = performanceService.now();
+					performanceService.processExecutionData(trackName, start, end, args);
+				} else {
+					resolvedPromise
+					.then(() => {
+						end = performanceService.now();
+						performanceService.processExecutionData(trackName, start, end, args);
+					})
+					.catch((err) => {
+						end = performanceService.now();
+						performanceService.processExecutionData(trackName, start, end, args);
+					});
+				}
+
+				return result;
+			}
+		};
+		descriptor.value = tempObject[originalMethod.name];
```

Comment:
This looks a little bit risky, we need to cover it with unit tests for not hiding exceptions, results, not affecting the call stack and so on.

Reply:
Will add tests for the decorator once the implementation is final.